cmd/pod-scaler/admission.go

```diff
@@ -273,36 +273,34 @@ func preventUnschedulable(resources *corev1.ResourceRequirements, cpuCap int64,
 }
 
 func mutatePodResources(pod *corev1.Pod, server *resourceServer, mutateResourceLimits bool, cpuCap int64, memoryCap string, reporter results.PodScalerReporter, logger *logrus.Entry) {
-	for i := range pod.Spec.InitContainers {
-		meta := pod_scaler.MetadataFor(pod.ObjectMeta.Labels, pod.ObjectMeta.Name, pod.Spec.InitContainers[i].Name)
-		resources, recommendationExists := server.recommendedRequestFor(meta)
-		if recommendationExists {
-			logger.Debugf("recommendation exists for: %s", pod.Spec.InitContainers[i].Name)
-			useOursIfLarger(&resources, &pod.Spec.InitContainers[i].Resources, determineName(pod.Labels, pod.Name, pod.Spec.InitContainers[i].Name), reporter, logger)
-			if mutateResourceLimits {
-				reconcileLimits(&pod.Spec.InitContainers[i].Resources)
-			}
-		}
-		preventUnschedulable(&pod.Spec.InitContainers[i].Resources, cpuCap, memoryCap, logger)
-	}
-	for i := range pod.Spec.Containers {
-		meta := pod_scaler.MetadataFor(pod.ObjectMeta.Labels, pod.ObjectMeta.Name, pod.Spec.Containers[i].Name)
-		resources, recommendationExists := server.recommendedRequestFor(meta)
-		if recommendationExists {
-			logger.Debugf("recommendation exists for: %s", pod.Spec.Containers[i].Name)
-			useOursIfLarger(&resources, &pod.Spec.Containers[i].Resources, determineName(pod.Labels, pod.Name, pod.Spec.Containers[i].Name), reporter, logger)
-			if mutateResourceLimits {
-				reconcileLimits(&pod.Spec.Containers[i].Resources)
+	mutateResources := func(containers []corev1.Container) {
+		for i := range containers {
+			meta := pod_scaler.MetadataFor(pod.ObjectMeta.Labels, pod.ObjectMeta.Name, containers[i].Name)
+			resources, recommendationExists := server.recommendedRequestFor(meta)
+			if recommendationExists {
+				logger.Debugf("recommendation exists for: %s", containers[i].Name)
+				useOursIfLarger(&resources, &containers[i].Resources, fmt.Sprintf("%s-%s", pod.Name, containers[i].Name), determineWorkloadType(pod.Annotations, pod.Labels), reporter, logger)
+				if mutateResourceLimits {
+					reconcileLimits(&containers[i].Resources)
+				}
 			}
+			preventUnschedulable(&containers[i].Resources, cpuCap, memoryCap, logger)
 		}
-		preventUnschedulable(&pod.Spec.Containers[i].Resources, cpuCap, memoryCap, logger)
 	}
+	mutateResources(pod.Spec.InitContainers)
+	mutateResources(pod.Spec.Containers)
 }
 
-// determineName returns the string that will be used in Prometheus metrics to identify the workload
-func determineName(podLabels map[string]string, podName, containerName string) string {
-	if value, exists := podLabels["prow.k8s.io/job"]; exists {
-		return value
+// determineWorkloadType returns the workflow type
+func determineWorkloadType(annotations, labels map[string]string) string {
+	if _, isBuildPod := annotations[buildv1.BuildLabel]; isBuildPod {
+		return "build"
+	}
+	if _, isProwjob := labels["prow.k8s.io/job"]; isProwjob {
+		return "prowjob"
+	}
+	if _, isRehearsal := labels[rehearse.Label]; isRehearsal {
```

Comment:
I am curious why you decided to call out rehearsals specifically, they are just another prowjob. Will this logic ever even be reached?

If we do want to call them out this will need to be before the check for prowjobs.

Reply:
Oh I didn't know rehearsals were a type of prowjob. In that case, we probably don't need this, I just found how to check for the rehearsals in the code above and thought it could be helpful.

Reply:
Yeah, lets not call out rehearsals here.